test/lit/passes/global-effects.wast

```diff
@@ -219,6 +219,53 @@
     )
   )
 
+  ;; WITHOUT:      (func $return-call-throw-and-catch (type $0)
+  ;; WITHOUT-NEXT:  (return_call $throw)
+  ;; WITHOUT-NEXT: )
+  ;; INCLUDE:      (func $return-call-throw-and-catch (type $0)
+  ;; INCLUDE-NEXT:  (return_call $throw)
+  ;; INCLUDE-NEXT: )
+  (func $return-call-throw-and-catch
+    (try
+      (do
+        ;; This call cannot be optimized out, as the target throws. However, the
+        ;; surrounding try-catch can be removed even without global effects
+        ;; because the throw from the return_call is never observed by this
+        ;; try-catch.
+        (return_call $throw)
+      )
+      (catch_all)
+    )
+  )
```

Comment:
@kripken, the part where we do not include the `throws` effect on the return_call itself is important for this test case. If we treat the `return_call` as throwing, we cannot optimize away the surrounding try-catch.

Reply:
Makes sense. But for correctness I think we need to fix the issue I mentioned.

I admit it seems theoretical atm, but e.g. `ignoreControlFlowTransfers()` will remove `branchesOut` which you mentioned as saving us from misoptimizing in that case. Now, `ignoreControlFlowTransfers()` is only used in one place that looks at more than a single instruction, so it ends up ok (it's not looking at the entire function), but this seems brittle.

Reply:
How would you fix the issue without regressing reasonable optimizations?

Reply:
I'm not sure yet, to be honest - there is definitely something awkward here in how this class doesn't quite match up to what we need it to do, given how `return_call` actually works. I'll think about this some more.